### PR TITLE
drivers: counter: Add native_posix counter hardware model and driver

### PR DESCRIPTION
**Pull request by FilipZajdel**
Adds native_posix hw counter model and the counter driver.
Functionality is needed by software which is tested
on native_posix and has dependency on counter.
Hardware model was developed similarly to HW timer model.
The counter driver wraps HW counter functions and exposes
basic functionalities: starting, stopping, setting and cancelling
single channel alarms.
Code was tested against: tests/drivers/counter/counter_basic_api.

Signed-off-by: Filip Zajdel <filip.zajdel@nordicsemi.no>

### CODE REVIEW

**Review by aescolar — DISMISSED** [review]
This is in the right direction

File: drivers/counter/counter_native_posix.c
Comment (FilipZajdel) [reply drivers/counter/counter_native_posix.c]: Done. Thanks.

File: boards/posix/native_posix/hw_counter.c
Comment (FilipZajdel) [reply boards/posix/native_posix/hw_counter.c:38]: Good point. Fixed. Thank you.

File: boards/posix/native_posix/hw_counter.c
Comment (FilipZajdel) [reply boards/posix/native_posix/hw_counter.c]: It's removed. Thanks.

File: boards/posix/native_posix/hw_counter.c
Comment (FilipZajdel) [reply boards/posix/native_posix/hw_counter.c]: Fixed. Thanks.

File: boards/posix/native_posix/hw_counter.c
Comment (FilipZajdel) [reply boards/posix/native_posix/hw_counter.c]: No reason. Thank you for mentioning it. Fixed.

File: boards/posix/native_posix/hw_counter.c
Comment (FilipZajdel) [reply boards/posix/native_posix/hw_counter.c]: You're right. This check is not very sophisticated. Fixed.

File: drivers/counter/counter_native_posix.c
Comment (FilipZajdel) [reply drivers/counter/counter_native_posix.c]: Solved. Thank you.

File: drivers/counter/counter_native_posix.c
Comment (FilipZajdel) [reply drivers/counter/counter_native_posix.c]: Thank you. Solved by introducing a variable indicating whether interrupt is pending.

**Comment by FilipZajdel** [comment]
@aescolar Thank you for the review. I fixed all the mentioned issues. Does it seem correct now?

**Comment by aescolar** [comment]
@nordic-krch Note that as maintainer for the counter your approval is also required before this could be merged.

File: drivers/counter/Kconfig.native_posix
Comment (aescolar) [inline drivers/counter/Kconfig.native_posix:9]: Better name it something like COUNTER_NATIVE_POSIX_FREQUENCY (or so, otherwise the option name is too generic)

**Review by aescolar — CHANGES_REQUESTED** [review]
Much better

File: boards/posix/native_posix/hw_counter.c
Comment (FilipZajdel) [reply boards/posix/native_posix/hw_counter.c]: Good point, I have renamed that. It's more consistent now.

File: drivers/counter/counter_native_posix.c
Comment (aescolar) [inline drivers/counter/counter_native_posix.c:40]: IRQ_CONNECT(COUNTER_EVENT_IRQ, 2, counter_isr, 0, 0);
NULL => 0 (that last field is flags (causes a compile warning))

File: drivers/counter/Kconfig.native_posix
Comment (FilipZajdel) [reply drivers/counter/Kconfig.native_posix]: Thank you. Done.

File: drivers/counter/counter_native_posix.c
Comment (FilipZajdel) [reply drivers/counter/counter_native_posix.c]: Fixed. Moreover I have #defined the other arguments to get rid of magic numbers.

**Review by aescolar — APPROVED** [review]

File: drivers/counter/counter_native_posix.c
Comment (nordic-krch) [inline drivers/counter/counter_native_posix.c:109]: just noticed that it is not documented in the API but other implementations returns -EBUSY if channel is already configured and did not yet expired.

**Review by nordic-krch — APPROVED** [review]
looks good. Just a minor comment about set alarm functionality.

File: drivers/counter/counter_native_posix.c
Comment (FilipZajdel) [reply drivers/counter/counter_native_posix.c:109]: So, would it conventionally be better to force user to cancel pending alarm (by returning -EBUSY) before setting it up again on a given channel? Did I understand it correctly?

**Pull request merged**
